### ISSUE TITLE
win_updates: category=security failing.

### ISSUE THREAD

**Issue by sonjz** [ansible/ansible-modules-extras#275]
Hi,

I've run `win_chocolatey: name=PSWindowsUpdate state=present` as specified beforehand, and here's my task to perform Windows Updates.

```
    - name: "Perform Windows Updates"
      win_updates: category=security
```

Appears to take time to download updates and then execution failes:

```
TASK: [Perform Windows Updates] ***********************************************
<10.1.12.97>
<10.1.12.97>
<10.1.12.97> category=security
<10.1.12.97>
<10.1.12.97>
<10.1.12.97>
<10.1.12.97>
<10.1.12.97> (offset=0 size=2028)
<10.1.12.97> (offset=2028 size=2028)
<10.1.12.97> (offset=4056 size=2028)
<10.1.12.97> (offset=6084 size=1570)
<10.1.12.97>
<10.1.12.97> (offset=0 size=24)
<10.1.12.97>
<10.1.12.97>
Traceback (most recent call last):
  File "/usr/lib/pymodules/python2.7/ansible/runner/connection_plugins/winrm.py", line 149, in exec_command
    result = self._winrm_exec(cmd_parts[0], cmd_parts[1:], from_exec=True)
  File "/usr/lib/pymodules/python2.7/ansible/runner/connection_plugins/winrm.py", line 120, in _winrm_exec
    vvvv('WINRM RESULT %r' % response, host=self.host)
  File "/usr/lib/pymodules/python2.7/ansible/callbacks.py", line 187, in vvvv
    return verbose(msg, host=host, caplevel=3)
  File "/usr/lib/pymodules/python2.7/ansible/callbacks.py", line 190, in verbose
    msg = utils.sanitize_output(msg)
  File "/usr/lib/pymodules/python2.7/ansible/utils/__init__.py", line 1002, in sanitize_output
    parts = parse_kv(str)
  File "/usr/lib/pymodules/python2.7/ansible/utils/__init__.py", line 793, in parse_kv
    vargs = split_args(args)
  File "/usr/lib/pymodules/python2.7/ansible/module_utils/splitter.py", line 185, in split_args
    raise Exception("error while splitting arguments, either an unbalanced jinja2 block or quotes")
Exception: error while splitting arguments, either an unbalanced jinja2 block or quotes
fatal: [10.1.12.97] => failed to exec cmd PowerShell -NoProfile -NonInteractive -EncodedCommand UABvAHcAZQByAFMAaABlAGwAbAAgAC0ATgBvAFAAcgBvAGYAaQBsAGUAIAAtAE4AbwBuAEkAbgB0AGUAcgBhAGMAdABpAHYAZQAgAC0ARQB4AGUAYwB1AHQAaQBvAG4AUABvAGwAaQBjAHkAIABVAG4AcgBlAHMAdAByAGkAYwB0AGUAZAAgAC0ARgBpAGwAZQAgACIAQwA6AFwAVQBzAGUAcgBzAFwAQQBkAG0AaQBuAGkAcwB0AHIAYQB0AG8AcgBcAEEAcABwAEQAYQB0AGEAXABMAG8AYwBhAGwAXABUAGUAbQBwAFwAYQBuAHMAaQBiAGwAZQAtAHQAbQBwAC0AMQA0ADIANAA1ADkANwAxADYAMQAuADIANAAtADEAMwA3ADUAMwAzADUANQAzADUAMgA5ADQAOAA1AFwAXAB3AGkAbgBfAHUAcABkAGEAdABlAHMALgBwAHMAMQAiACAAIgBDADoAXABVAHMAZQByAHMAXABBAGQAbQBpAG4AaQBzAHQAcgBhAHQAbwByAFwAQQBwAHAARABhAHQAYQBcAEwAbwBjAGEAbABcAFQAZQBtAHAAXABhAG4AcwBpAGIAbABlAC0AdABtAHAALQAxADQAMgA0ADUAOQA3ADEANgAxAC4AMgA0AC0AMQAzADcANQAzADMANQA1ADMANQAyADkANAA4ADUAXABcAGEAcgBnAHUAbQBlAG4AdABzACIAOwAgAFIAZQBtAG8AdgBlAC0ASQB0AGUAbQAgACIAQwA6AFwAVQBzAGUAcgBzAFwAQQBkAG0AaQBuAGkAcwB0AHIAYQB0AG8AcgBcAEEAcABwAEQAYQB0AGEAXABMAG8AYwBhAGwAXABUAGUAbQBwAFwAYQBuAHMAaQBiAGwAZQAtAHQAbQBwAC0AMQA0ADIANAA1ADkANwAxADYAMQAuADIANAAtADEAMwA3ADUAMwAzADUANQAzADUAMgA5ADQAOAA1AFwAIgAgAC0ARgBvAHIAYwBlACAALQBSAGUAYwB1AHIAcwBlADsA

FATAL: all hosts have already failed -- aborting

PLAY RECAP ********************************************************************
           to retry, use: --limit @/root/sql_server.retry

10.1.12.97                 : ok=4    changed=2    unreachable=1    failed=0
```

**Comment by sonjz** [comment]
I've done a bit of investigating and running the `PSWindowsUpdate` from my own `ps1` script and found that there is a common issue running COM objects utilized by WinRM.

test.ps1

```

  Write-Host "Test" ;
  # NOTE: this executes and should report the follow (you'll notice the WebProxy should say System.__ComObject but doesn't when run from WinRM)
  # PS C:\>   $objSession = New-Object -ComObject "Microsoft.Update.Session" ;
  # PS C:\>   $objSession ;
  # ClientApplicationID                                ReadOnly WebProxy                                         UserLocale
  # -------------------                                -------- --------                                         ----------
  #                                                       False System.__ComObject                                     1033
  $objSession = New-Object -ComObject "Microsoft.Update.Session" ;
  $objSession ;

  # NOTE: this is where we get the traceback error as an exception is thrown for uninstantiated $objSession
  $downloader = $objSession.CreateUpdateDownloader() ;
  $downloader ;

```

base.yml

```
---
    - name: "Wait for WinRM"
      local_action:
        module: wait_for
        host: "{{ inventory_hostname }}"
        port: 5986
        timeout: 900
    #- pause: seconds=15

    # NOTE: win_updates not working as expected, run manually
    - name: "Perform Windows Updates"
      script: test.ps1
      register: shell_output
    - debug: var=shell_output
```

Here is a a message thread on the same issue (different COM object):
https://social.technet.microsoft.com/Forums/windowsserver/en-US/55394ff1-f9f0-49a5-bc62-2c78a2107661/problems-with-psremoting?forum=winserverpowershell

tl;dr
Appears to be permission issues to use the COM object, which is part of PSWindowsUpdate package.

**Comment by sonjz** [comment]
Some other links that suggest this approach has no joy:

https://social.technet.microsoft.com/Forums/scriptcenter/en-US/930f41ee-8a46-46ce-837c-2032fd99325a/invokewuinstall-remote-host?forum=ITCG#c5c5e6bf-78ac-4934-be70-851972e6d00b

```
.DESCRIPTION

Use Invoke-WUInstall to invoke Windows Update install remotly. It Based on TaskScheduler because

CreateUpdateDownloader() and CreateUpdateInstaller() methods can't be called from a remote computer - E_ACCESSDENIED.

Note:

Because we do not have the ability to interact, is recommended use -AcceptAll with WUInstall filters in script block
```

http://serverfault.com/questions/336705/issues-with-patching-servers-remotely-using-winrm-and-microsoft-update-session

In what scenario was this tested and working?

**Comment by sonjz** [comment]
Here's my solution for the issue.  Instead of trying to run anything directly from the PSWindowsUpdate module, I created a ps1 that registers a scheduled job, and ask ansible to poll for a file to mark its completion of its task.  In my usage, I also `cat` the contents of the log so my ansible log records it.

Here is the task being called in its entirety:

```
    - name: "Enable Powershell script execution"
      raw: powershell -ExecutionPolicy ByPass -Command Set-ExecutionPolicy ByPass -force

    # NOTE: win_updates not working as expected, created scheduled job fix
    - name: "Invoke Windows Update"
      script: assets/Invoke-WindowsUpdate.ps1

    # NOTE: can we please get wait_for path in Windows !
    #       retry for 60 min, every 15 sec, 3600/15 = 240 retries
    - name: "Waiting for Windows Update to complete"
      win_stat: path=C:\PSWindowsUpdate.DONE
      register: shell_output
      until: shell_output.stat.exists == true
      retries: 240
      delay: 15
    - raw: powershell -ExecutionPolicy ByPass -Command Get-Content C:\PSWindowsUpdate.DONE

    - name: "Disabled Powershell script execution"
      raw: powershell -Command Set-ExecutionPolicy Restricted -force
```

Invoke-Windows-Update.ps1

```
# author: Jason Huggins (git @sonjz)
# date created: 20150223
# Windows Update using PSWindowsUpdate but with a scheduled job.
# In your ansible task, add the following:
#    # NOTE: win_updates not working as expected, created scheduled job fix
#    - name: "Perform Windows Updates"
#      script: assets/Invoke-WindowsUpdate.ps1
#
#    # NOTE: can we please get wait_for path in Windows !
#    #       retry for 60 min, every 15 sec, 3600/15 = 240 retries
#    - name: "Waiting for Windows Update to complete"
#      win_stat: path=C:\PSWindowsUpdate.DONE
#      register: shell_output
#      until: shell_output.stat.exists == true
#      retries: 240
#      delay: 15
#    - raw: powershell -ExecutionPolicy ByPass -Command Get-Content C:\PSWindowsUpdate.DONE
#
# Uses PSWindowsUpdate:
# - https://gallery.technet.microsoft.com/scriptcenter/2d191bcd-3308-4edd-9de2-88dff796b0bc
#
# Solution drawn from:
# - https://gallery.technet.microsoft.com/scriptcenter/Force-Install-Updates-on-ef821f9a
#
# Why not win_Update:
# There is an issue with using Microsoft.Update.Session::CreateUpdateDownloader remotely
# additionally, PSWindowsUpdate::Invoke-WUInstall appeared to have a similar issue remotely
function Invoke-WindowsUpdate {

  $invokeScript = {
    $baseFile = "C:\PSWindowsUpdate";

    if (Test-Path("$baseFile.DONE")) {
      Move-Item "$baseFile.DONE" "$baseFile.$((Get-Date -Format O) -replace ':','').BAK" ;
    }
    Import-Module PSWindowsUpdate ;
    Get-WUInstall -AcceptAll -IgnoreReboot | Out-File "$baseFile.PROCESSING" ;

    # NOTE: personally removing KB2966828", otherwise server can't webpi .NET 3.5 if desired later
    wusa.exe /uninstall /kb:2966828 /quiet /log /norestart >> "$baseFile.PROCESSING" ;
    Move-Item "$baseFile.PROCESSING" "$baseFile.DONE" ;
  } ;

  # remove the job (if exists), create a new one.
  # NOTE: this will clobber any previous job
  $jobName = "Invoke-WindowsUpdate" ;
  $result = Get-ScheduledJob | Where { $_.Name -eq "$jobName" } ;
  if ($result) {
    Unregister-ScheduledJob -Name "$jobName" ;
  }
  Register-ScheduledJob -Name "$jobName" -RunNow -ScriptBlock $invokeScript ;
}

Invoke-WindowsUpdate ;
```

EDIT: wrapped task with Set-ExecutionPolicy

**Comment by adematte** [comment]
just to let you know that we're facing the same issue and also had to rely on a scheduled job to work around it for now. If it proves stable enough, maybe I will look into submitting a pull request for the win_updates extra module.

**Comment by sonjz** [comment]
Here's a slightly better solution where the "wait" step occurs in the ps1.
For an actual ansible module, you could probably just place the DONE file in the tmp folder or use an environment variable (but I'm not sure its limit).

Invoke-Windows-Update.ps1

```
# author: Jason Huggins (git @sonjz)
# date created: 20150223
# Windows Update using PSWindowsUpdate but with a scheduled job.
# In your ansible task, add the following:
#    # NOTE: win_updates not working as expected, created scheduled job fix
#    - name: "Perform Windows Updates"
#      script: assets/Invoke-WindowsUpdate.ps1
#
# Uses PSWindowsUpdate:
# - https://gallery.technet.microsoft.com/scriptcenter/2d191bcd-3308-4edd-9de2-88dff796b0bc
#
# Solution drawn from:
# - https://gallery.technet.microsoft.com/scriptcenter/Force-Install-Updates-on-ef821f9a
#
# Why not win_Update:
# There is an issue with using Microsoft.Update.Session::CreateUpdateDownloader remotely
# additionally, PSWindowsUpdate::Invoke-WUInstall appeared to have a similar issue remotely
function Invoke-WindowsUpdate {

  $invokeScript = {
    $baseFile = "C:\Invoke-WindowsUpdate";

    if (Test-Path("$baseFile.DONE")) {
      Move-Item "$baseFile.DONE" "$baseFile.$((Get-Date -Format O) -replace ':','').BAK" ;
    }
    Import-Module PSWindowsUpdate ;
    Get-WUInstall -AcceptAll -IgnoreReboot | Out-File "$baseFile.PROCESSING" ;

    # NOTE: personally removing KB2966828", otherwise server can't webpi .NET 3.5 if desired later
    wusa.exe /uninstall /kb:2966828 /quiet /log /norestart >> "$baseFile.PROCESSING" ;
    Move-Item "$baseFile.PROCESSING" "$baseFile.DONE" ;
  } ;

  # remove the job (if exists), create a new one.
  # NOTE: this will clobber any previous job
  $jobName = "Invoke-WindowsUpdate" ;
  $result = Get-ScheduledJob | Where { $_.Name -eq "$jobName" } ;
  if ($result) {
    Unregister-ScheduledJob -Name "$jobName" ;
  }
  Register-ScheduledJob -Name "$jobName" -RunNow -ScriptBlock $invokeScript ;
}

Invoke-WindowsUpdate ;


# monitor for DONE file and report
$doneFile =  "C:\Invoke-WindowsUpdate.DONE" ;
Start-Sleep -s 5 ;
while(!(Test-Path $doneFile)) {
  Start-Sleep -s 5 ;
}
Get-Content $doneFile ;

```

**Comment by dgoade** [comment]
@sonjz Thanks for your research on this. It took me most of the day but I'm really glad I found this issue and will be trying your workaround myself.

**Comment by petemounce** [comment]
@sonjz This was tested and working against eu-west-1 ami-5d62ff2a, a Windows 2012r2 base AMI owned by Amazon (at the time, it was a different AMI, but same series). It was launched with user data like:

``` powershell
<powershell>
Set-ExecutionPolicy RemoteSigned
icm $executioncontext.InvokeCommand.NewScriptBlock((New-Object Net.WebClient).DownloadString('https://s3-eu-west-1.amazonaws.com/public.prs.je-labs.com/vagrant/001_prep_for_provisioning.ps1'))
icm $executioncontext.InvokeCommand.NewScriptBlock((New-Object Net.WebClient).DownloadString('https://s3-eu-west-1.amazonaws.com/public.prs.je-labs.com/vagrant/ConfigureRemotingForAnsible.ps1'))
</powershell>
```

then had some things installed
- chocolatey
- ruby + ruby devkit
- sysinternals
- `User-Interfaces-Infra` windows feature _un_installed, then a reboot with `win_reboot`
- `pswindowsupdate`, then critical updates then security updates.

I didn't see errors. Excellent research and work you've done here!

It looks like your background-job is definitely the way to go. (I'm creating a `win_ec2config` module that will try that approach to run the sysprep step, which will shut down the instance - if it's run inline, then the winrm command will error because it will never return).

**Comment by nitzmahone** [comment]
win_updates was completely rewritten in 2.0; closing
